main/lsp/request_dispatch.cc

```diff
@@ -55,33 +46,31 @@ LSPResult LSPLoop::processRequestInternal(unique_ptr<core::GlobalState> gs, cons
         auto &params = msg.asNotification().params;
         if (method == LSPMethod::SorbetWorkspaceEdit) {
             // Note: We increment `lsp.messages.processed` when the original requests were merged into this one.
-            auto &editParams = get<unique_ptr<SorbetWorkspaceEditParams>>(params);
-            const u4 end = editParams->updates.versionEnd;
-            const u4 start = editParams->updates.versionStart;
-            // Versions are sequential and wrap around. Use them to figure out how many edits are contained within this
-            // update.
-            const u4 merged = min(end - start, 0xFFFFFFFF - start + end);
-            auto run = runTypechecking(move(gs), move(editParams->updates));
-            // Only report stats if the edit was committed.
-            if (!run.canceled) {
-                prodCategoryCounterInc("lsp.messages.processed", "sorbet/workspaceEdit");
-                prodCategoryCounterAdd("lsp.messages.processed", "sorbet/mergedEdits", merged);
-            }
-            return commitTypecheckRun(move(run));
+            shared_ptr<SorbetWorkspaceEditParams> editParams = move(get<unique_ptr<SorbetWorkspaceEditParams>>(params));
+            // Since std::function is copyable, we have to promote captured unique_ptrs into shared_ptrs.
+            // TODO(jvilk): Switch to asyncRun once I sort out how this interplays with cancelable slow path.
```

Comment:
I can't have this use `asyncRun` right now, as it will break cancelable slow path. Specifically, cancelable slow path _currently_ assumes that, if a slow path run is running, the coordinator thread is blocked waiting for it -- so no other edits could have snuck in.

However, if the slow path is now async, a fast path edit may be waiting to typecheck.

I haven't yet figured out how to reconcile these issues.

Reply:
I think I can make this async once preemption works. Here's an example scenario:

* Slow path edit comes in (parse error), starts typechecking.
* Fast path edit comes in (file still has parse error), preempts thread and typechecks.
* New edit comes in (fixes parse error), we detect that going from edit currently running on slow path -> this new edit takes fast path, so we cancel slow path and proceed with typechecking new update.